analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs

```diff
@@ -0,0 +1,143 @@
+                }
+            }, SymbolKind.NamedType);
+        });
+
+    private static void ProcessControllerMethods(SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        if (context.SemanticModel.GetDeclaredSymbol(context.Node) is IMethodSymbol method
+            && method.IsControllerMethod())
+        {
+            var actionParameters = method.Parameters
+                .Select(x => x.Type)
+                .OfType<INamedTypeSymbol>()
+                .Where(x => !IgnoreType(x))
+                .SelectMany(x => RelatedTypesToExamine(x))
+                .Distinct();
+            foreach (var parameter in actionParameters)
+            {
+                CheckInvalidProperties(parameter, context, examinedTypes);
+            }
+        }
+    }
+
+    private static void CheckInvalidProperties(INamedTypeSymbol parameterType, SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        var invalidProperties = GetAllDeclaredProperties(parameterType, examinedTypes)
+                .Where(x => !CanBeNull(x.Type)
+                    && !x.HasAttribute(KnownType.System_ComponentModel_DataAnnotations_RequiredAttribute)
+                    && !x.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ModelBinding_Validation_ValidateNeverAttribute));
+        foreach (var property in invalidProperties)
+        {
+            var propertySyntax = property.DeclaringSyntaxReferences[0].GetSyntax();
+            context.ReportIssue(Rule, propertySyntax.GetLocation());
+        }
+    }
+
+    private static bool IgnoreType(ITypeSymbol type) =>
+        type.IsAny(IgnoredTypes)
+        || (!type.DeclaringSyntaxReferences.Any()
+            && !type.Is(KnownType.System_Collections_Generic_IEnumerable_T)
+            && !type.Implements(KnownType.System_Collections_Generic_IEnumerable_T));
+
+    private static bool CanBeNull(ITypeSymbol type) =>
+        type is ITypeParameterSymbol { HasValueTypeConstraint: false }
+        || (type.IsReferenceType && type.NullableAnnotation() != NullableAnnotation.NotAnnotated)
+        || type.IsNullableValueType();
+
+    private static IEnumerable<IPropertySymbol> GetAllDeclaredProperties(ITypeSymbol type, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        var allProperties = Enumerable.Empty<IPropertySymbol>();
+        if (type is INamedTypeSymbol namedType && examinedTypes.TryAdd(namedType, true))
+        {
+            var properties = namedType.GetMembers()
+                .OfType<IPropertySymbol>()
+                .Where(x => x.GetEffectiveAccessibility() == Accessibility.Public
+                    && x.SetMethod?.DeclaredAccessibility is Accessibility.Public);
+            foreach (var property in properties)
+            {
+                allProperties = allProperties.Append(property);
```

Comment:
The property is only of interest if `DeclaringSyntaxReferences` references a non-autogenerated syntax tree. Adding it here unconditionally may cause a out of bound exception in line 99.